feat(ci): publish platform specific binaries

numToStr opened the pull request:
Summary of changes:
- Now platform-specific artifacts will be uploaded on each run
- Once a tag is released then it will populate the [releases](https://github.com/sumneko/lua-language-server/releases) page with platform-specific assets. See https://github.com/numToStr/lua-language-server/releases

---

Some questions I have

- Platform suffix in the build path (ie. `bin/{Linux,Windows,macOS}`) bothers me a little as this is also reflected inside the release assets. Also, IMO these are kinda redundant now. So, Are you ok with removing the suffix? So that assets could just have binary w/o any directories.

```lua
-- Current
"bin/Linux" -- Bin dir
"lua-language-server-linux/linux/lua-language-server" -- inside asset
--     ^                     ^
-- asset name         platform suffix    

-- Wanted
"bin" -- bin dir
"lua-language-server-linux/lua-language-server" -- asset
-- or
"lua-language-server-linux/bin/lua-language-server" -- alternative structure
```

- ~We can probably merge this https://github.com/sumneko/lua-language-server/blob/master/.github/workflows/checkPR.yml workflow inside build workflow. What do you think?~ Done ;)

Closes #849

numToStr commented:
I think I might also need to update the wiki :)

clason commented:
Note that the binaries are not enough to have a functional installation; you also need the runtime scripts: https://github.com/sumneko/lua-language-server/issues/849#issuecomment-987969795

clason commented:
And +1 on flattening the hierarchy for the binary; that would make packaging for Homebrew easier!

numToStr commented:
> 5. `copy (recursively) main.lua, debugger.lua, locale, meta, and script to build`

@clason Are you talking about this?

clason commented:
yes

numToStr commented:
Ok. I'll give it a shot.

clason commented:
Also, a standard archive with the full (including submodules) sources for each release would be nice for Homebrew ;)

numToStr commented:
Isn't the source zip automatically uploaded in the release?

clason commented:
I don't think that includes the submodules?

numToStr commented:
> I don't think that includes the submodules? At least it doesn't for the tags.

Just checked. It doesn't even have the `.git` folder XD

> (Not at all a dealbreaker, but if that's an easy add to your action, it'd be nice.)

I suppose I could do that :)

clason commented:
I don't want the `.git` folder, just the pulled dependencies ;)

numToStr commented:
> I don't want the `.git` folder, just the pulled dependencies ;)

Got it. I'll probably tackle this tomorrow :p

actboy168 commented:
>
> Platform suffix in the build path (ie. `bin/{Linux,Windows,macOS}`) bothers me a little as this is also reflected inside the release assets. Also, IMO these are kinda redundant now. So, Are you ok with removing the suffix? So that assets could just have binary w/o any directories.
> 


I guess the bin directory contains platform suffix, because the release version of VSCode needs to contain all platform binaries. 

But VSCode already supports platform specific extension. If lua-language-server can support this feature, it will no longer be necessary to include multiple platform binaries in one release.

numToStr commented:
> If lua-language-server can support this feature, it will no longer be necessary to include multiple platform binaries in one release. 

Yes, the goal is to release platform specific binary packages so that anyone can play with lua-server as quickly as possible.

> But VSCode already supports platform specific extension. 

I don't know anything about vscode extensions but if it supports platform specific extensions then we are in good position.

sumneko commented:
I will try to modify the VSCode release process and remove the <platform> folder

numToStr commented:
@sumneko Should I remove the `<platform>` folder here? I suppose this is the line https://github.com/sumneko/lua-language-server/blob/247377652b61fd21f0a16ff8fd15bc8c54e3e080/make.lua#L5

sumneko commented:
> @sumneko Should I remove the `<platform>` folder here? I suppose this is the line
> 
> https://github.com/sumneko/lua-language-server/blob/247377652b61fd21f0a16ff8fd15bc8c54e3e080/make.lua#L5

There are still some problems with modifying the directory structure, we need to keep it unchanged first.

numToStr commented:
@clason I've added the runtime files inside the release assets and also added the tagged version in the assets name. See https://github.com/numToStr/lua-language-server/releases/tag/0.0.0-pre.9

numToStr commented:
> There are still some problems with modifying the directory structure, we need to keep it unchanged first.

@sumneko Understood. I'll let you update the directory structure :)

numToStr commented:
Now the submodules asset is also ready :) https://github.com/numToStr/lua-language-server/releases/tag/0.0.0-pre.11

numToStr commented:
@sumneko I am mostly done. If you have any suggestions please let me know :)

sumneko commented:
@numToStr I am a bit busy during this time, so I can't actively follow up on this matter.

If I understand correctly, the current effect is that a complete package will be uploaded to the `Artifacts` every time the `master` is pushed.
When the `tag` is pushed, the complete package will be published to `release`.

If this is the case, there is no problem.

numToStr commented:
> @numToStr I am a bit busy during this time, so I can't actively follow up on this matter.

No worries. There is no rush to merge this :)

> If I understand correctly, the current effect is that a complete package will be uploaded to the `Artifacts` every time the `master` is pushed. When the `tag` is pushed, the complete package will be published to `release`.
> 
> If this is the case, there is no problem.

Yes, exactly.

---

Also, If you want I can also merge the `checkPR` workflow in this :)

sumneko commented:
> Also, If you want I can also merge the `checkPR` workflow in this :)

I didn't know much about `workflow` before, I just divided the different functions into different files according to my idea.
If the mainstream design is written in the same file, then you can help me merge it together.

numToStr commented:
> If the mainstream design is written in the same file, then you can help me merge it together.

Yeah sure. We can definitely merge them as previously both workflow were effectively the same.

clason commented:
I think the idea was to remove the platform-specific directories (`macOS` etc.), not the `bin` one?

clason commented:
If the platform name must remain included for technical reasons, could it at least be changed to the output of `uname -s` (which would be `Linux` on Linux and `Darwin` on macOS`)?

numToStr commented:
Yes, `bin` is already there inside the assets and I didn't remove anything as sumneko raised his concerns before.

![image](https://user-images.githubusercontent.com/24727447/145954822-511beef5-6d12-475f-9248-bba72f826e43.png)

I don't know whether this folder structure will work or not, if not just let me know :)

clason commented:
Sorry, I shouldn't have commented here as this is about the current commits (which seem to be going in a different direction than your PR)!

numToStr commented:
> (The `uname` suggestion still applies, though.)

Yes, I am also on board with that but @sumneko will be the better person to do it.

(Although I prefer to just remove the platform name ;))

clason commented:
Agreed; the point was _not_ to just flatten the hierarchy (turning `bin/<platform>/` into `bin-<platform>` just breaks every script out there without any actual benefit...)

sumneko commented:
I just changed the folder structure to this:
![image](https://user-images.githubusercontent.com/5213431/145955851-5300d097-92a2-4040-a6ef-d7e48919ee44.png)

My ultimate goal is indeed to keep only one `bin`, without the need to include platform information, but there are currently some obstacles:
1. VSCode's extension is released from my local computer to all platforms, so it must contain all the binary files of the platform at the same time. Need to be published separately to each platform (`vsce publish --target <target>`).
2. My local platform is Windows, so the Linux binary will lose its running permissions when it is released. Need to be published through workflow instead.
3. How to publish workflow to the VSCode market through my token still needs investigation.
4. This extension will also be published to https://open-vsx.org at the same time (`ovsx publish [extension.vsix] --pat <token>`). This extension market does not support sub-platforms, and the token is clear text through command line parameters, maybe I have to temporarily abandon this market after modifying the release process.

clason commented:
@sumneko I see, thank you for explaining! Does VS Code require the platform names to be the way they are now (`macOS` in particular)?

sumneko commented:
> @sumneko I see, thank you for explaining! Does VS Code require the platform names to be the way they are now (`macOS` in particular)?

No. It is just because reason 1 above.

sumneko commented:
Maybe we can generate 4 versions in workflow at the same time:
1. Windows(win32) with `bin`
2. Linux(linux) with `bin`
3. macOS(darwin) with `bin`
4. AllInOne with `bin-Windows`, `bin-Linux` and `bin-macOS`

This can meet the needs of non-VSCode first, and I will slowly change the process of VSCode later.

numToStr commented:
I think this will be the best course of action

(but I am wondering if publishing a platform-specific extension is not too much then I could also help with that ;))

sumneko commented:
> I think this will be the best course of action
> 
> (but I am wondering if publishing a platform-specific extension is not too much then I could also help with that ;))

What is too much mean?

numToStr commented:
> What is too much mean?

Like setting up action and configuring tokens etc.

> Maybe we can generate 4 versions in workflow at the same time:
> 
> 1. Windows(win32) with `bin`
> 2. Linux(linux) with `bin`
> 3. macOS(darwin) with `bin`
> 4. AllInOne with `bin-Windows`, `bin-Linux` and `bin-macOS`
> 

I am also fine with this suggestion. Should I update the workflow then?

actboy168 commented:
One PR should only do one thing. Putting too many changes in a PR will only make things get out of control.

sumneko commented:
Maybe open a new clear PR?

numToStr commented:
> Maybe open a new clear PR?

I think this PR is doing what I meant for it to do ie. platform-specific binaries. It is just the changes related to directory ie. `bin-{Windows,macOS,Linux}`. If you can revert that and maybe we can later do that with the all-in-one package for vscode-extension :)

sumneko commented:
> 

You could just overwrite it

numToStr commented:
Yes, doable. But now I have to store `Windows`, `macOS`, and `Linux` to correctly pick the build folder.

sumneko commented:
Which version do you want to revert to? `./bin/Linux/lua-language-server` or `./Linux/lua-language-server` or other?

numToStr commented:
> Which version do you want to revert to? `./bin/Linux/lua-language-server` or `./Linux/lua-language-server` or other?

To this `./bin/Linux/lua-language-server` as It was before. Thanks :)

sumneko commented:
It should be reverted.

numToStr commented:
Thanks, @sumneko. I'll quickly rebase my PR.

numToStr commented:
@sumneko I think I am finished with this PR. You can check the latest [run](https://github.com/numToStr/lua-language-server/actions/runs/1578093771) and [release assets](https://github.com/numToStr/lua-language-server/releases/tag/0.0.3)

sumneko commented:
What is `submodules.zip` ?

numToStr commented:
> What is `submodules.zip`?

This contains all the submodules sources for homebrew.

sumneko commented:
I think the `./build` folder should be excluded

numToStr commented:
Done.

sumneko commented:
`./log` should also be excluded.
I still not understand why it is needed, the `./3rd/lovr` is so large

numToStr commented:
> `./log` should also be excluded. I still not understand why it is needed, the `./3rd/lovr-api` is so large

@clason asked me for the submodule asset for homebrew distribution. I could remove it if you want.

sumneko commented:
> > `./log` should also be excluded. I still not understand why it is needed, the `./3rd/lovr-api` is so large
> 
> @clason asked me for the submodule asset for homebrew distribution. I could remove it if you want.

I used to think of `homebrew` as `darwin`, although they didn't look same at all.
If it is necessary for package management, there is no problem (I am not paying for the file size and bandwidth anyway)

sumneko commented:
It seems good, thank you!

clason commented:
> > > `./log` should also be excluded. I still not understand why it is needed, the `./3rd/lovr-api` is so large
> > 
> > 
> > @clason asked me for the submodule asset for homebrew distribution. I could remove it if you want.
> 
> I used to think of `homebrew` as `darwin`, although they didn't look same at all. If it is necessary for package management, there is no problem (I am not paying for the file size and bandwidth anyway)

I just need the `3rd/luamake` and `3rd/bee` subdirectories for building. The idea is to be able to just download the release and call `luamake rebuild` without needing a `git submodule init update`.

> I used to think of homebrew as darwin, although they didn't look same at all.
If it is necessary for package management, there is no problem (I am not paying for the file size and bandwidth anyway)

Homebrew used to be macOS only, but they have joined with the `linuxbrew` project, so now they build everything for macOS _and_ Linux from the same formula (hence needing to distinguish the platforms there).

clason commented:
To make clear, it would simplify the homebrew formula if the `v2.6.0.tar.gz` release tarball included the submodules needed for building it, but it is not _necessary_ (homebrew can clone a repo and update submodules as part of the build pipeline, like it does for the current formula I submitted). So if that is at all awkward to do here, just don't worry about including submodules.
